web/packages/teleport/src/Roles/RoleEditor/standardmodel.ts

```diff
@@ -603,3 +587,124 @@ export function hasModifiedFields(
+    case 'kube_cluster':
+      return runRules(spec, kubernetesValidationRules);
+    case 'node':
+      return runRules(spec, serverValidationRules);
+    case 'app':
+      return runRules(spec, appSpecValidationRules);
+    case 'db':
+      return runRules(spec, databaseSpecValidationRules);
+    case 'windows_desktop':
+      return runRules(spec, windowsDesktopSpecValidationRules);
+    default:
+      kind satisfies never;
+  }
+}
+
+export type AccessSpecValidationResult =
+  | ServerSpecValidationResult
+  | KubernetesSpecValidationResult
+  | AppSpecValidationResult
+  | DatabaseSpecValidationResult
+  | WindowsDesktopSpecValidationResult;
+
+const validKubernetesResource = (res: KubernetesResourceModel) => () => {
+  const name = requiredField(
+    'Resource name is required, use "*" for any resource'
+  )(res.name)();
+  const namespace = kubernetesClusterWideResourceKinds.includes(res.kind.value)
+    ? { valid: true }
+    : requiredField('Namespace is required for resources of this kind')(
+        res.namespace
+      )();
+  return {
+    valid: name.valid && namespace.valid,
+    name,
+    namespace,
+  };
+};
+export type KubernetesResourceValidationResult = {
+  name: ValidationResult;
+  namespace: ValidationResult;
+};
+
+const kubernetesValidationRules = {
+  labels: nonEmptyLabels,
+  resources: arrayOf(validKubernetesResource),
+};
+export type KubernetesSpecValidationResult = RuleSetValidationResult<
+  typeof kubernetesValidationRules
+>;
+
+const noWildcard = (message: string) => (value: string) => () => {
+  const valid = value !== '*';
+  return { valid, message: valid ? '' : message };
+};
+
+const noWildcardOptions = (message: string) => (options: Option[]) => () => {
+  const valid = options.every(o => o.value !== '*');
+  return { valid, message: valid ? '' : message };
+};
+
+const serverValidationRules = {
+  labels: nonEmptyLabels,
+  logins: noWildcardOptions('Wildcard is not allowed in logins'),
+};
+export type ServerSpecValidationResult = RuleSetValidationResult<
+  typeof serverValidationRules
+>;
+
+const appSpecValidationRules = {
+  labels: nonEmptyLabels,
+  awsRoleARNs: arrayOf(noWildcard('Wildcard is not allowed in AWS role ARNs')),
+  azureIdentities: arrayOf(
+    noWildcard('Wildcard is not allowed in Azure identities')
+  ),
+  gcpServiceAccounts: arrayOf(
+    noWildcard('Wildcard is not allowed in GCP service accounts')
+  ),
+};
+export type AppSpecValidationResult = RuleSetValidationResult<
+  typeof appSpecValidationRules
+>;
+
+const databaseSpecValidationRules = {
+  labels: nonEmptyLabels,
+  roles: noWildcardOptions('Wildcard is not allowed in database roles'),
+};
+export type DatabaseSpecValidationResult = RuleSetValidationResult<
+  typeof databaseSpecValidationRules
+>;
+
+const windowsDesktopSpecValidationRules = {
+  labels: nonEmptyLabels,
+};
+export type WindowsDesktopSpecValidationResult = RuleSetValidationResult<
+  typeof windowsDesktopSpecValidationRules
+>;
```

Comment:
nit: all the rules functions in this file would have a better home in a separate `rule.ts` file

Reply:
Yes, I agree. The only reason I didn't make it is that I'm constantly in progress with more PRs, so after the version cut, I'm planning to split both the StandardEditor file, as well as this one.

Reply:
(Oh, wait. This is actually new code. It can be easily extracted.)